aea/aea_builder.py

```diff
@@ -0,0 +1,746 @@
+                )
+            )
+
+    def _check_can_add(self, configuration: ComponentConfiguration) -> None:
+        """
+        Check if the component can be added, given its configuration.
+
+        :param configuration: the configuration of the component.
+        :return: None
+        :raises ValueError: if the component is not present.
+        """
+        self._check_configuration_not_already_added(configuration)
+        self._check_package_dependencies(configuration)
+
+    def set_name(self, name: str) -> "AEABuilder":
+        """
+        Set the name of the agent.
+
+        :param name: the name of the agent.
+        """
+        self._name = name
+        return self
+
+    def set_default_connection(self, public_id: PublicId):
+        """
+        Set the default connection.
+
+        :param public_id: the public id of the default connection package.
+        :return: None
+        """
+        self._default_connection = public_id
+
+    def add_private_key(
+        self, identifier: str, private_key_path: PathLike
+    ) -> "AEABuilder":
+        """
+        Add a private key path.
+
+        :param identifier: the identifier for that private key path.
+        :param private_key_path: path to the private key file.
+        """
+        self._private_key_paths[identifier] = str(private_key_path)
+        return self
+
+    def remove_private_key(self, identifier: str) -> "AEABuilder":
+        """
+        Remove a private key path by identifier, if present.
+
+        :param identifier: the identifier of the private key.
+
+        """
+        self._private_key_paths.pop(identifier, None)
+        return self
+
+    @property
+    def private_key_paths(self) -> Dict[str, str]:
+        """Get the private key paths."""
+        return self._private_key_paths
+
+    def add_ledger_api_config(self, identifier: str, config: Dict):
+        """Add a configuration for a ledger API to be supported by the agent."""
+        self._ledger_apis_configs[identifier] = config
+
+    def remove_ledger_api_config(self, identifier: str):
+        """Remove a ledger API configuration."""
+        self._ledger_apis_configs.pop(identifier, None)
+
+    @property
+    def ledger_apis_config(self) -> Dict[str, Dict[str, Union[str, int]]]:
+        """Get the ledger api configurations."""
+        return self._ledger_apis_configs
+
+    def set_default_ledger_api_config(self, default: str):
+        """Set a default ledger API to use."""
+        self._default_ledger = default
+
+    def add_component(
+        self, component_type: ComponentType, directory: PathLike
+    ) -> "AEABuilder":
+        """
+        Add a component, given its type and the directory.
+
+        :param component_type: the component type.
+        :param directory: the directory path.
+        :raises ValueError: if a component is already registered with the same component id.
+        """
```

Comment:
In summary:
- we first load the configuration in the directory, and do preliminary checks (e.g. right aea version, consistent fingerprints etc.)
- Then, we load the package by including all the dependencies added so far to the agent builder (`manager.dependencies()`). They are handled with a context manager that removes the additional packages imported by the loaded module. For instance, if `connection.py` imports some protocol it will be removed after loading. Other library packages like `requests` will be left in `sys.modules` (they could be removed too anyway).
-